### ISSUE TITLE
summary v2 create_file_writer() behavior leads to incorrect resource deletion

### ISSUE THREAD

**Issue by nfelt** [tensorflow/tensorflow#25707]
This is a tracking bug for the known issue that the V2 summary API's `create_file_writer()` logic behaves unexpectedly in regards to the underlying resource lifecycle management.  This typically manifests as users seeing errors like `tensorflow.python.framework.errors_impl.NotFoundError: Resource localhost/logdir:./log/N10tensorflow22SummaryWriterInterfaceE does not exist. [Op:WriteScalarSummary] name: epoch_loss/`

Users can encounter this in 1.x when using `tf.contrib.summary.create_file_writer()` in eager mode, or in 2.0 as `tf.summary.create_file_writer()`, or via wrapper APIs like the Keras TensorBoard callback.

Example issues: #24632, #25524

The root cause is that `create_file_writer(logdir)` returns a SummaryWriter python object backed by a C++ SummaryWriterInterface resource, but the mapping is not 1:1, which violates other assumptions about resource management under eager mode.  In particular, two SummaryWriter instances created for the same logdir by default will both attempt to reference the same named resource.  (Sharing by logdir was intended to be a mechanism for using one eventfile per logdir akin to `tf.summary.FileWriterCache` in 1.x, which makes TensorBoard better behaved.)

This is a problem when one of the SummaryWriters is deleted (aka loses its last remaining reference), because it will unconditionally attempt to delete the named resource it was created with.  If a resource with that same name is referenced by any other SummaryWriters, those ones will now emit the above `NotFoundError` when attempting to use them.  Note that this happens even if the deleted SummaryWriter was properly `close()`'d before the new one was created, because the logdir name sharing means that it new writer's resource has the same name as the old one.

The correct fix is changing the resource names to be unique so that the mapping from SummaryWriter object <-> SummaryWriterInterface resource is 1:1 as designed.  This will likely require other workarounds for the one-eventfile-per-logdir problem.

Minimal repro:
```python
import tempfile
import tensorflow as tf

if tf.__version__.startswith('2.'):
  create_file_writer = tf.summary.create_file_writer
else:
  tf.enable_eager_execution()
  create_file_writer = tf.contrib.summary.create_file_writer

dir = tempfile.mkdtemp('shared-writer-bug')
w = create_file_writer(dir)
w.close()  # succeeds, deletes old resource
# 1) new SummaryWriter created w/ new resource (but same name)
# 2) reassigning w deletes old SummaryWriter along w/ new resource (BUG)
w = create_file_writer(dir)
w.close()  # raises NotFoundError; new resource has been incorrectly deleted
```

**Comment by ataiya** [comment]
Is there a workaround while the bug is being resolved?

**Comment by nfelt** [comment]
Reopening to track propagating this fix into the TF 2.0 Keras TensorBoard callback, which will fix the user issues reported in the original description above.

**Comment by wchargin** [comment]
Fixed by 6068ebcbb82a19a40fa7cdb36da51b62d2a11a5c.

**Comment by wchargin** [comment]
Closed by 059ea3ba68db861e40d750eba688281011d2735f.

**Comment by yatbear** [comment]
Seeing a similar error in `scalars_demo`:
```
$ bazel run tensorboard/plugins/scalar:scalars_demo --test_output=all
INFO: Analyzed target //tensorboard/plugins/scalar:scalars_demo (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //tensorboard/plugins/scalar:scalars_demo up-to-date:
  bazel-bin/tensorboard/plugins/scalar/scalars_demo
INFO: Elapsed time: 0.208s, Critical Path: 0.01s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
2022-08-11 15:59:10.131986: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2022-08-11 15:59:10.132032: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2022-08-11 15:59:12.548022: E tensorflow/stream_executor/cuda/cuda_driver.cc:271] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2022-08-11 15:59:12.548134: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (themachine.c.googlers.com): /proc/driver/nvidia/version does not exist
Saving output to /tmp/scalars_demo.
--- Running: t0=270,tA=270,kH=0.001
2022-08-11 15:59:12.550387: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
--- Running: t0=270,tA=270,kH=0.005
2022-08-11 15:59:13.761329: W tensorflow/core/framework/op_kernel.cc:1745] OP_REQUIRES failed at summary_kernels.cc:138 : NOT_FOUND: Resource localhost/_AnonymousVar0/N10tensorflow22SummaryWriterInterfaceE does not exist.
Traceback (most recent call last):
  File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 164, in <module>
    app.run(main)
  File "/usr/local/google/home/ytjing/workspace/yatbear/tensorboard/tf/lib/python3.10/site-packages/absl/app.py", line 312, in run
    _run_main(main, args)
  File "/usr/local/google/home/ytjing/workspace/yatbear/tensorboard/tf/lib/python3.10/site-packages/absl/app.py", line 258, in _run_main
    sys.exit(main(argv))
  File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 142, in main
    run_all(LOGDIR, verbose=True)
  File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 133, in run_all
    run(
  File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 106, in run
    one_step(temp, ambient_temperature, heat_coefficient, step)
  File "/usr/local/google/home/ytjing/workspace/yatbear/tensorboard/tf/lib/python3.10/site-packages/tensorflow/python/util/traceback_utils.py", line 153, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/usr/local/google/home/ytjing/workspace/yatbear/tensorboard/tf/lib/python3.10/site-packages/tensorflow/python/eager/execute.py", line 54, in quick_execute
    tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
tensorflow.python.framework.errors_impl.NotFoundError: Graph execution error:

Detected at node 'temperature/current/write_summary' defined at (most recent call last):
    File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 164, in <module>
      app.run(main)
    File "/usr/local/google/home/ytjing/workspace/yatbear/tensorboard/tf/lib/python3.10/site-packages/absl/app.py", line 312, in run
      _run_main(main, args)
    File "/usr/local/google/home/ytjing/workspace/yatbear/tensorboard/tf/lib/python3.10/site-packages/absl/app.py", line 258, in _run_main
      sys.exit(main(argv))
    File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 142, in main
      run_all(LOGDIR, verbose=True)
    File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 133, in run_all
      run(
    File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 106, in run
      one_step(temp, ambient_temperature, heat_coefficient, step)
    File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/scalars_demo.py", line 46, in one_step
      tf.summary.scalar(
    File "/usr/local/google/home/ytjing/.cache/bazel/_bazel_ytjing/0a906c509cf65041941e0f57a3d21533/execroot/org_tensorflow_tensorboard/bazel-out/k8-fastbuild/bin/tensorboard/plugins/scalar/scalars_demo.runfiles/org_tensorflow_tensorboard/tensorboard/plugins/scalar/summary_v2.py", line 93, in scalar
      metadata=summary_metadata,
Node: 'temperature/current/write_summary'
Resource localhost/_AnonymousVar0/N10tensorflow22SummaryWriterInterfaceE does not exist.
	 [[{{node temperature/current/write_summary}}]] [Op:__inference_one_step_62]
```

Googlers, see b/239742608.
